src/network/hobbits/hobbitsConnectHandler.ts

```diff
@@ -0,0 +1,266 @@
+
+
+/**
+ * The NetworkRpc module controls network-level resources and concerns of p2p connections
+ */
+export class HobbitsConnectHandler extends EventEmitter {
+  /**
+   * dials in progress
+   */
+  private wipDials: Set<string>;
+  /**
+   * Connected peers mapped to peer id
+   */
+  private peers: Map<string, Peer>;
+  /**
+   * Milliseconds to request timeout
+   */
+  private requestTimeout: number;
+
+  /**
+   * Active outgoing request ids mapped to methods
+   */
+  private requests: Record<RequestId, Method>;
+
+  /**
+   * Active incoming request ids mapped to peers and methods
+   */
+  private responses: Record<RequestId, {peer: Peer; method: Method}>;
+
+  private server: net.Server;
+  public logger: ILogger;
+  private bootnodes: string[];
+
+  public constructor(bootnodes: string[], logger: ILogger) {
+    super();
+    this.bootnodes = bootnodes;
+    this.logger = logger;
+    this.requestTimeout = 5000;
+    this.requests = {};
+    this.responses = {};
+    this.peers = new Map<string, Peer>();
+    // Create a server for accepting remote connection
+    this.server = net.createServer();
+  }
+
+  public addPeer(hobbitsUri: HobbitsUri): Peer {
+    let peerId = hobbitsUri.toUri();
+    let peer = this.peers.get(peerId);
+    if (!peer) {
+      peer = new Peer(hobbitsUri, this);
+      this.peers.set(peerId, peer);
+      // rpc peer connect
+      this.emit("peer:connect", hobbitsUri);
+    }
+    return peer;
+  }
+  public removePeer(hobbitsUri: HobbitsUri): void {
+    let peerId = hobbitsUri.toUri();
+    const peer = this.peers.get(peerId);
+    if (peer) {
+      peer.close();
+      this.peers.delete(peerId);
+      this.emit("peer:disconnect", hobbitsUri);
+    }
+  }
+  public getPeers(): HobbitsUri[] {
+    return Array.from(this.peers.values()).map((p) => p.hobbitsUri);
+  }
+
+  public hasPeer(hobbitsUri: HobbitsUri): boolean {
+    const peerId = hobbitsUri.toUri();
+    return this.peers.has(peerId);
+  }
+
+  /*public async onConnection(protocol: string, conn: Connection): Promise<void> {
+    let peerInfo;
+    while (!peerInfo) {
+      peerInfo = await promisify(conn.getPeerInfo.bind(conn))();
+    }
+    const peer = this.addPeer(peerInfo);
+    peer.setConnection(conn, true);
+  }*/
+
+  public onConnectionEnd(hobbitsUri: HobbitsUri): void {
+    this.removePeer(hobbitsUri);
+  }
+
+  public async dialForRpc(hobbitsUri: HobbitsUri): Promise<void> {
+    const peerId = hobbitsUri.toUri();
+    if (this.peers.has(peerId)) {
+      return;
+    }
+    if (this.wipDials.has(peerId)) {
+      return;
+    }
+    this.wipDials.add(peerId);
+    try {
+      const peer = this.addPeer(hobbitsUri);
+      peer.connect();
+    } catch (e) {
+      this.logger.error(e.stack);
+    }
+    this.wipDials.delete(peerId);
+  }
+
+  public async sendRequest<T extends ResponseBody>(hobbitsUri: HobbitsUri, type: ProtocolType, method: Method, body: RequestBody): Promise<T> {
+    const peerId = hobbitsUri.toUri();
```

Comment:
@wemeetagain `hobbits` has 3 kinds of messages (RPC, GOSSIP, PING) protocol. So `sendRequest` needs to know the `ProtocolType`. Can we modify it to add a `type` parameter?